File: c_src/flash.c
/*
  This file is a part of JOP, the Java Optimized Processor

  Copyright (C) 2001-2008, Martin Schoeberl (martin@jopdesign.com)

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
	flash.c

	read ttf and write in upper half of flash for config

	Mem.java must be running on JOP


**********
	read file for flash (Motorola S records) and write flash
	attached to JOP3

	Author: Martin Schoeberl martin@good-ear.com

*/

#include <windows.h>
#include <stdio.h>
#include <string.h>

#define FLASH_SIZE 0x20000

main(int argc, char *argv[]) {

	char c;
	int i, j;
	char inbuf[100];
	int cnt;

	unsigned char mem[2*FLASH_SIZE];

	FILE *fp;
	char buf[1000];
	unsigned char *cp;
	int s_type, s_len, s_a, s_d;
	int val;
	int data, adr;
	int end;

	DCB dcb;
	HANDLE hCom;
	DWORD dwError;
	BOOL fSuccess;


	if (argc!=2) {
		printf("usage: flash file\n");
		exit(-1);
	}
	if ((fp=fopen(argv[1],"r"))==NULL) {
		printf("Error opening %s\n", argv[1]);
		exit(-1);
	}

	hCom = CreateFile("COM1",
		GENERIC_READ | GENERIC_WRITE,
		0,    /* comm devices must be opened w/exclusive-access */
		NULL, /* no security attrs */
		OPEN_EXISTING, /* comm devices must use OPEN_EXISTING */
		0,    /* not overlapped I/O */
		NULL  /* hTemplate must be NULL for comm devices */
		);

	if (hCom == INVALID_HANDLE_VALUE) {	/* try COM2 */
		hCom = CreateFile("COM2", GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
	}

	if (hCom == INVALID_HANDLE_VALUE) {
		dwError = GetLastError();

		printf("shit CreateFile\n");
		exit(-1);
	}

/*
 * Omit the call to SetupComm to use the default queue sizes.
 * Get the current configuration.
 */

	fSuccess = GetCommState(hCom, &dcb);

	if (!fSuccess) {
		printf("shit GetCommState\n");
		exit(-1);
	}

	dcb.BaudRate = 115200;
	dcb.ByteSize = 8;
	dcb.Parity = NOPARITY;
	dcb.StopBits = ONESTOPBIT;

	/* use RTS/CTS handshake */
	dcb.fOutxCtsFlow = TRUE;
	dcb.fRtsControl = RTS_CONTROL_ENABLE;

	dcb.fOutxDsrFlow = FALSE;
	dcb.fDtrControl = DTR_CONTROL_DISABLE;
	dcb.fDsrSensitivity = FALSE;

	fSuccess = SetCommState(hCom, &dcb);

	if (!fSuccess) {
		printf("shit SetCommState\n");
		exit(-1);
	}

//	config via Max+ (jopbb)
//	system("acex ttf\\flash.ttf");

	for (i=0; i<2*FLASH_SIZE; ++i) {
		mem[i] = 0xff;
	}

/******** read S format ***********
	for( ; cp = fgets(buf, sizeof(buf), fp); cp!=NULL) {
		if (*cp++!='S') { printf("no 'S'\n"); exit(-1); }
		s_type = *cp++;
		sscanf(cp,"%2x", &s_len);
		s_len -= 1;
		cp +=2;
		switch (s_type) {
			case '0':
			case '1':
			case '9':
				sscanf(cp,"%4x", &s_a);
				cp += 4;
				s_len -= 2;
				break;
			case '2':
				sscanf(cp,"%6x", &s_a);
				cp += 6;
				s_len -= 3;
				break;
			case '3':
				sscanf(cp,"%8x", &s_a);
				cp += 8;
				s_len -= 4;
				break;
			case '5':
				break;
			case '7':
				break;
			default:
				printf("not a valid S-record\n");
				exit(-1);
				break;
		}
		switch (s_type) {
			case '0':
				for (i=0; i<s_len; ++i) {
					sscanf(cp, "%2x", &s_d);
					printf("%c", s_d);
					cp += 2;
				}
				printf("\n");
				break;
			case '1':
			case '2':
			case '3':
//				printf("%04x ", s_a);
				for (i=0; i<s_len; ++i) {
					sscanf(cp, "%2x", &s_d);
//					printf("%02x ", s_d);
					mem[s_a++] = s_d;
					cp += 2;
				}
//				printf("\n");
				break;
			case '9':
				printf("start address: %04x\n", s_a);
				break;
			default:
				printf("typ %c ???\n", s_type);
				break;
		}
	}
**************** end read S format **********/


/********* read ttf ***********/


	for ( i=0; fscanf(fp, "%d", &val)==1; ++i) {
		mem[i] = val;
		fscanf(fp, ",");
	}
	fclose(fp);
	end = i+128;

/********* end read ttf ***********/

//
//	program flash
//
	for (i=0; i<end; i+=128) {
		sprintf(buf, "a%x\n", i);
		wr(hCom, buf);
		for (j=0; j<128; ++j) {
			adr = i+j;
			data = mem[adr];
			sprintf(buf, "d%xw", data);
			wr(hCom, buf);
			printf("\r");
		}
		sprintf(buf, "a%xp", adr);
		wr(hCom, buf);
		for (;;) {
			wr(hCom, "r");
			ReadFile(hCom, inbuf, 1, &cnt, NULL);
			ReadFile(hCom, inbuf+1, 1, &cnt, NULL);
			inbuf[2] = 0;
			sscanf(inbuf, "%x", &val);
			printf("%02x", val);
			if (val==data) {
				break;
			}
		}
		printf("\n");
	}
	printf("%d %x data\n", i, i);
//
//	compare
//
	wr(hCom, "a0 ");
	for (i=0; i<end; ++i) {
		wr(hCom, "i");
		printf("%d\r", i);
		ReadFile(hCom, inbuf, 1, &cnt, NULL);
		ReadFile(hCom, inbuf+1, 1, &cnt, NULL);
		inbuf[2] = 0;
		sscanf(inbuf, "%x", &val);
		if (val!=mem[i]) {
			printf("wrong data at %x: %02x-%02x\n", i, mem[i], val);
		}
	}

	fclose(fp);
	CloseHandle(hCom);
}

#define MAX_CNT 10

int wr(HANDLE hCom, char *buf) {

	char inbuf[100];
	int i;
	int cnt;

	inbuf[0] = 0;
	for (i=0; i<strlen(buf); ++i) {
		WriteFile(hCom, buf+i, 1, &cnt, NULL);
		ReadFile(hCom, inbuf, 1, &cnt, NULL); printf("%c", *inbuf);
	}
}

File: java/pc/src/udp/Flash.java
/*
  This file is part of JOP, the Java Optimized Processor (http://www.jopdesign.com/)

  Copyright (C) 2001-2008, Martin Schoeberl (martin@jopdesign.com)

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/**
*	Flash programming via TFTP.
*
*	For new jopcore board with 512 KB flash.
*
*	three file types:
*
*		.ttf	ACEX config start at 0x60000
*				Cyclone config start at 0x40000
*		.bin	Java files start at 0x10000		(history of BB project)
*		.html	start at 0x00000
*		.class	start at 0x30000 (for Applet Tal.class)
*
*	Flash overview (in 64 KB sectors):
*
*		0x00000 :	Java program
*		0x10000 :	HTML or longer Java program
*		0x20000 :	user data, Streckendaten in oebb BG263
*		0x30000 :	Applet for TAL, bgid in oebb
*		0x40000 :	CYC config
*		0x50000 :	CYC config
*		0x60000 :	ACX, CYC config
*		0x70000 :	ACX, CYC config
*
*	Changelog:
*		2008-05-22	ACEX configuration disabled as we use a compressed bitstream for
*					the Cyclone and size detection does not work.
*/
package udp;
import java.io.*;
import java.util.*;

public class Flash extends FlashConst {

	protected static final int FLASH_SIZE = 0x80000;
	protected static final int SECTOR_SIZE = 0x10000;
	protected static final int SECTOR_MASKx = 0xf0000;
	protected static final int SECTOR_SHIFT = 16;

	protected static final int MAX_ACEX = 32768*3;	// 96kB (acex 1k50)
	protected static final int MAX_MEM = 65536*4;		// 256kB (Cyclone)
	protected static final int MAX_JAVA = 16384*4*2;	// max. 128KB
	protected static final int MAX_HTML = 1024;		// max. 1KB

	protected static final int START_JAVA = 0x00000;
	protected static final int START_HTML = 0x10000;
	protected static final int START_DATA = 0x20000;
	protected static final int START_CONFIG = 0x30000;
	protected static final int START_ACX_TTF = 0x60000;
	protected static final int START_CYC_TTF = 0x40000;
	
	protected int start;

	protected String fname;

	protected byte[] mem;
	protected int len;
	private Tftp tftp;

	public Flash(Tftp t, String fname) {

		tftp = t;
		this.fname = fname;
		mem = new byte[MAX_MEM];
		for (int i=0; i<MAX_MEM; ++i) mem[i] = (byte) 0xff;
	}

	public void read() {

		String line;
		StringTokenizer st;
		int pos;
		boolean isTtf = false;
		boolean isJava = false;
		boolean isHtml = false;
		boolean isAppl = false;
		int byteCnt = 0;

		len = 0;
		try {
			File f = new File(fname);
			String s = f.getName();
			pos = s.indexOf('.');
			if (pos==-1) {
				System.out.println("wrong filename");
				System.exit(-1);
			}
			s = s.substring(pos);
			if (s.equals(".jop") || s.equals(".bin")) {
				isJava = true;
				start= START_JAVA;
			} else if (s.equals(".html")) {
				isHtml = true;
				start= START_HTML;
			} else if (s.equals(".class")) {
				isAppl = true;
				start= START_CONFIG;
			} else if (s.equals(".ttf")) {
				isTtf = true;
				// TODO: should be changed to an option
//				start= START_ACX_TTF;		// assume a file for Jopcore (ACEX)
				start = START_CYC_TTF;
			} else {
				System.out.println("wrong file type: '"+s+"'");
				System.exit(-1);
			}

			if (isJava || isTtf) {

				BufferedReader in = new BufferedReader(new FileReader(f));

				while ((line = in.readLine()) != null) {

					if ((pos = line.indexOf('/'))!=-1) {
						st = new StringTokenizer(line.substring(0, pos), " \t\n\r\f,");
					} else {
						st = new StringTokenizer(line, " \t\n\r\f,");
					}
					while (st.hasMoreTokens()) {
						int val = Integer.parseInt(st.nextToken());
	
						if (isJava) {
							mem[len++] = (byte) (val>>>24);
							mem[len++] = (byte) (val>>>16);
							mem[len++] = (byte) (val>>>8);
							mem[len++] = (byte) val;
						} else {
							mem[len++] = (byte) val;
						}
	
						if (len==MAX_MEM) {
							System.out.println("too many words");
							System.exit(-1);
						}
						if (isJava && len==MAX_JAVA) {
							System.out.println("too many words for this Flash");
							System.exit(-1);
						}
					}
				}
				in.close();

//				if (len>MAX_ACEX) {							// file is big, so it must be for Cyclone
//					start= START_CYC_TTF;					// change start address
//				}

			} else {					// 'binary' file

				InputStream in = new FileInputStream(f);
				len = in.read(mem);
				if (in.available()!=0) {
					System.out.println("file to long");
					System.exit(-1);
				}
				if (isHtml) {
					mem[len] = 0;			// end character for html files
					++len;
				} else if (isAppl) {		// length of class file in first two bytes
					for (int i=len-1; i>=0; --i) {
						mem[i+2] = mem[i];
					}
					mem[0] = (byte) (len>>>8);
					mem[1] = (byte) len;
					len += 2;
				}
			}

		} catch (IOException e) {
			System.out.println(e.getMessage());
			System.exit(-1);
		}
		System.out.println(len+" bytes to program");
	}

	public void program() {

		int i, j;

		byte[] buf = new byte[SECTOR_SIZE];
		byte[] inbuf = new byte[SECTOR_SIZE];


		for (i=0; i<len; i+=SECTOR_SIZE) {

			int slen = SECTOR_SIZE;
			if (len-i<slen) slen = len-i;

			System.out.println("programming");
			for (j=0; j<slen ; ++j) {
				buf[j] = mem[i+j];
			}
			byte s = (byte) ('0'+((i+start)>>SECTOR_SHIFT));
System.out.println("sector "+(s-'0'));

			try {
				if (!tftp.write((byte) 'f', (byte) s, buf, slen)) {
					System.out.println();
					System.out.println("programming error");
					System.exit(-1);
				}
				System.out.println();
				System.out.println("compare");
				if (tftp.read((byte) 'f', s, inbuf)!=SECTOR_SIZE) {
					System.out.println();
					System.out.println("read error");
					System.exit(-1);
				}
			} catch (Exception e) {
				System.out.println(e);
				System.exit(-1);
			}

			for (j=0; j<slen; ++j) {
				if (buf[j] != inbuf[j]) {
					System.out.println("wrong data: "+(i+j)+" "+buf[j]+" "+inbuf[j]);
					System.exit(-1);
				}
			}
		}
		System.out.println("programming ok");
	}

	public static void main (String[] args) {

		if (args.length < 2) {
			System.out.println("usage: java Flash file host");
			System.exit(-1);
		}

		Flash fl = new Flash(new Tftp(args[1]), args[0]);
		fl.read();
		fl.program();
	}
}
